crates/compiler/solve/src/solve.rs

```diff
@@ -1600,6 +1686,48 @@ fn solve(
                 if should_check_exhaustiveness {
                     use roc_can::exhaustive::{check, ExhaustiveSummary};
 
+                    // If the condition type likely comes from an positive-position value (e.g. a
+                    // literal or a return type), rather than an input position, we employ the
+                    // heuristic that the positive-position value would only need to be open if the
+                    // branches of the `when` constrained them as open. To avoid suggesting
+                    // catch-all branches, now mark the condition type as closed, so that we only
+                    // show the variants that explicitly not matched.
+                    //
+                    // We avoid this heuristic if the condition type likely comes from a negative
+                    // position, e.g. a function parameter, since in that case if the condition
+                    // type is open, we definitely want to show the catch-all branch as necessary.
+                    //
+                    // For example:
+                    //
+                    //   x : [A, B, C]
+                    //
+                    //   when x is
+                    //      A -> ..
+                    //      B -> ..
+                    //
+                    // This is checked as "almost equal" and hence exhaustiveness-checked with
+                    // [A, B] compared to [A, B, C]*. However, we really want to compare against
+                    // [A, B, C] (notice the closed union), so we optimistically close the
+                    // condition type here.
+                    //
+                    // On the other hand, in a case like
+                    //
+                    //   f : [A, B, C]* -> ..
+                    //   f = \x -> when x is
+                    //     A -> ..
+                    //     B -> ..
+                    //
+                    // we want to show `C` and/or `_` as necessary branches, so this heuristic is
+                    // not applied.
+                    //
+                    // In the above case, notice it would not be safe to apply this heuristic if
+                    // `C` was matched as well. Since the positive/negative value determination is
+                    // only an estimate, we also only apply this heursitic in the "almost equal"
+                    // case, when there was in fact a unification error.
+                    if cond_source_is_likely_positive_value && has_unification_error {
+                        close_pattern_matched_tag_unions(subs, real_var);
+                    }
```

Comment:
Should probably add a note that ideally we can get rid of this heuristic once we figure out https://roc.zulipchat.com/#narrow/stream/231635-compiler-development/topic/Modelling.20open.20tags